streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java

```diff
@@ -469,48 +465,76 @@ private void pipeRecord(final String topicName,
             validateSourceTopicNameRegexPattern(topicName);
         }
         final TopicPartition topicPartition = getTopicPartition(topicName);
-        if (topicPartition != null) {
-            final long offset = offsetsByTopicPartition.get(topicPartition).incrementAndGet() - 1;
-            task.addRecords(topicPartition, Collections.singleton(new ConsumerRecord<>(
-                    topicName,
-                    topicPartition.partition(),
-                    offset,
-                    timestamp,
-                    TimestampType.CREATE_TIME,
-                    (long) ConsumerRecord.NULL_CHECKSUM,
-                    key == null ? ConsumerRecord.NULL_SIZE : key.length,
-                    value == null ? ConsumerRecord.NULL_SIZE : value.length,
-                    key,
-                    value,
-                    headers)));
-
-            // Process the record ...
-            task.process(mockWallClockTime.milliseconds());
-            task.maybePunctuateStreamTime();
-            task.commit();
-            captureOutputRecords();
+
+        if (topicPartition == null) {
+            processGlobalRecord(topicName, timestamp, key, value, headers);
         } else {
-            final TopicPartition globalTopicPartition = globalPartitionsByTopic.get(topicName);
-            if (globalTopicPartition == null) {
-                throw new IllegalArgumentException("Unknown topic: " + topicName);
+            enqueueTaskRecord(topicName, topicPartition, timestamp, key, value, headers);
+            processAllProcessableRecords();
+        }
+    }
+
+    private void enqueueTaskRecord(final String topicName,
+                                   final TopicPartition topicPartition,
+                                   final Long timestamp,
+                                   final byte[] key,
+                                   final byte[] value,
+                                   final Headers headers) {
+        if (!internalTopologyBuilder.sourceTopicNames().isEmpty()) {
+            validateSourceTopicNameRegexPattern(topicName);
+        }
+        Objects.requireNonNull(topicPartition);
```

Comment:
Should we add the same check for `topicName`, `timestamp`, and `headers` ?

Reply:
I can; I only added this check because it was never possible to go into this logic before with a null topicPartition (because it was inside a condition block that `topicPartition != null`), but now that it's a separate method, future (buggy) code changes could result in passing a `null` here. We'd just get an NPE later on, but I wanted to check it up front on a line of its own to decrease debugging time.